### PR TITLE
Llama3 hybrid implementation using submeshes

### PR DESCRIPTION
**Pull request by ipotkonjak-tt**
### Problem description
Missing support for data / hybrid parallelism for Llama3 models.

### What's changed
Addition of hybrid parallelism within llama code base with concept of submeshes. Implementation is mainly based at the LlamaGenerator level. MeshDevice is partitioned into submeshes where each subset of devices has an independent model. Models remain implemented in the tensor parallel manner.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/13856034978) CI passes
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/runs/13856051711) CI passes
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/runs/13719186263) CI passes
- [x] [(Single-card) Nightly model and ttnn](https://github.com/tenstorrent/tt-metal/actions/runs/13902343318) CI passes
- [x] [(T3K) T3000 demo](https://github.com/tenstorrent/tt-metal/actions/runs/13905084535) CI passes
- [x] [(T3K) T3000 frequent](https://github.com/tenstorrent/tt-metal/actions/runs/13902293366) CI passes
- [x] [(T3K) T3000 unit](https://github.com/tenstorrent/tt-metal/actions/runs/13902299082) CI passes
- [x] New/Existing tests provide coverage for changes - added test cases to demo tests

### CODE REVIEW

File: models/demos/llama3/tt/model_config.py [hunk 1]
```diff
@@ -530,18 +530,18 @@ def __init__(
...
-                ttnn.ShardSpec(
-                    core_range_set_by_batch,
-                    [
-                        128,
-                        128,
-                    ],
-                    ttnn.ShardOrientation.ROW_MAJOR,
-                ),
-            )
+            # self.model_config["ROT_MAT_MEMCONFIG"] = ttnn.MemoryConfig(
```
Comment (yieldthought) [inline models/demos/llama3/tt/model_config.py:533]: Good find that this isn't used anywhere, just delete it instead of commenting out though

**Review by yieldthought — CHANGES_REQUESTED** [review]
Clean 👌

To do:
- [ ] Add at least one CI test that will exercise DP. I suggest adding a demo to the [t3k tests](tests/scripts/t3000/run_t3000_demo_tests.sh).

File: models/demos/llama3/tt/model_config.py (hunk 1 above)
Comment (ipotkonjak-tt) [reply models/demos/llama3/tt/model_config.py]: Unused code is removed

**Review by cfjchu — APPROVED** [review]

File: models/demos/llama3/demo/simple_text_demo.py [hunk 2]
```diff
@@ -155,7 +147,82 @@ def create_tt_model(
...
+    page_table = None
+    paged_attention_config = None
+
+    if use_paged_kv_cache:
+        paged_attention_config = PagedAttentionConfig(
+            block_size=page_params["page_block_size"],
+            max_num_blocks=page_params["page_max_num_blocks"],
+        )
+        # Implied shuffling of blocks
+        permutation = torch.randperm(paged_attention_config.max_num_blocks)
```
Comment (skhorasganiTT) [inline models/demos/llama3/demo/simple_text_demo.py:163]: According to this, `max_num_blocks` now represents the max blocks per dp group right? Can it be renamed to `max_num_blocks_per_dp` for clarity?

File: models/demos/llama3/demo/simple_vision_demo.py [hunk 3]
```diff
@@ -84,13 +116,21 @@ def create_multimodal_model(mesh_device, max_batch_size, max_seq_len, dtype=ttnn
     ],
     ids=["batch1-notrace", "batch1-trace", "batch32-trace", "batch4-trace-with-text-prompts"],
 )
+@pytest.mark.parametrize(
+    "data_parallel",
+    [
+        1,
+        # 4,
```
Comment (skhorasganiTT) [inline models/demos/llama3/demo/simple_vision_demo.py:123]: Should this DP test be uncommented?

File: models/demos/llama3/tt/generator_vllm.py [hunk 4]
```diff
@@ -118,12 +169,28 @@ def __init__(self, *args, **kwargs):
...
+                max_seq_len=max_seq_len,
+                use_paged_kv_cache=True,
+                checkpoint=state_dict,
+            )
+            model_args.append(model_args_i)
+            model.append(model_i)
+
         return cls(model, model_args, mesh_device)
 
     @property
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator_vllm.py:196]: Should be model_args[0] in `cache_path` and `max_cross_attn_tokens`?

File: models/demos/llama3/tt/generator_vllm.py [hunk 5]
```diff
@@ -191,14 +261,17 @@ def initialize_vllm_model(cls, hf_config, mesh_device, max_batch_size, n_layers=
...
 
     def decode_forward(self, *args, **kwargs):
         return super().decode_forward_text(*args, **kwargs)
 
+    def allocate_kv_cache(self, *args, **kwargs):
+        return allocate_kv_cache(*args, **kwargs)
+
 
 class TtQwen2ForCausalLM(LlamaGenerator):
     def __init__(self, *args, **kwargs):
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator_vllm.py:277]: Should be model_args[0] in cache_path?

File: models/demos/llama3/tt/generator_vllm.py [hunk 6]
```diff
@@ -118,12 +169,28 @@ def __init__(self, *args, **kwargs):
...
     @classmethod
     def initialize_vllm_model(cls, hf_config, mesh_device, max_batch_size):
         max_seq_len = 131072
-        model_args, model = create_multimodal_model(mesh_device, max_batch_size, max_seq_len, use_paged_kv_cache=True)
+
+        data_parallel, submesh_devices = generate_submeshes(mesh_device)
+
+        model_args = []
+        model = []
+
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator_vllm.py:182]: missing `state_dict=None` for first loop iter

File: models/demos/llama3/tt/generator_vllm.py [hunk 7]
```diff
@@ -20,6 +20,44 @@
 from vllm.model_executor.models.mllama import MLLAMA_IMAGE_TOKEN_ID, MLLAMA_IMAGE_TOKEN
 
 
+def generate_submeshes(mesh_device):
+    data_parallel = int(os.getenv("TT_DATA_PARALLEL", 1))
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator_vllm.py:24]: Could you replace this env var with a new arg like `tt_data_parallel` in the `initialize_vllm_model` class methods? I think it would be better than propagating an env var all the way here

File: models/demos/llama3/tt/generator_vllm.py [hunk 8]
```diff
@@ -20,6 +20,44 @@
...
+    if not isinstance(mesh_device, ttnn.MeshDevice) or data_parallel == 1:
+        return 1, [mesh_device]
+
+    num_devices = mesh_device.get_num_devices()
+    assert num_devices % data_parallel == 0, f"Unsupported device split: {num_devices} devices, {data_parallel} groups"
+
+    return data_parallel, mesh_device.create_submeshes(ttnn.MeshShape(1, num_devices // data_parallel))
+
+
+def allocate_kv_cache(kv_cache_shape, dtype, num_layers, mesh_device):
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator_vllm.py:35]: TODO (@ipotkonjak-tt and/or @skhorasganiTT) Modify KV creation in vLLM to use this function and test with DP

File: models/demos/llama3/demo/simple_text_demo.py [hunk 9]
```diff
@@ -262,6 +363,8 @@ def create_tt_model(
         "reasoning-1",  # reasoning
         "ci-1",  # CI batch 1
         "ci-32",  # CI batch 32
+        "batch-1-DP-4",  # DP 4 latency
+        "batch-1-DP-8",  # DP 8 latency
```
[the remark this answers is not in the document]
Comment (ipotkonjak-tt) [reply models/demos/llama3/demo/simple_text_demo.py:383]: Added a test for batch 32 and DP 4

File: models/demos/llama3/demo/simple_text_demo.py [hunk 10]
```diff
@@ -155,7 +147,82 @@ def create_tt_model(
...
+        mesh_device.create_submeshes(ttnn.MeshShape(1, num_devices // data_parallel))
+        if isinstance(mesh_device, ttnn.MeshDevice) and data_parallel > 1
+        else [mesh_device]
+    )
+    state_dict = None
+
+    # Hybrid requires a model per submesh
+    model_args = []
+    model = []
+    page_table = []
```
[the remark this answers is not in the document]
Comment (ipotkonjak-tt) [reply models/demos/llama3/demo/simple_text_demo.py]: Removed line

File: models/demos/llama3/tt/generator_vllm.py (hunk 5 above)
Comment (ipotkonjak-tt) [reply models/demos/llama3/tt/generator_vllm.py:279]: Yes, changed

File: models/demos/llama3/tt/generator_vllm.py (hunk 6 above)
Comment (ipotkonjak-tt) [reply models/demos/llama3/tt/generator_vllm.py:183]: Added line

File: models/demos/llama3/tt/generator_vllm.py (hunk 7 above)
Comment (ipotkonjak-tt) [reply models/demos/llama3/tt/generator_vllm.py]: Env var replaced with `tt_data_parallel` arg in `initialize_vllm_model` and `allocate_kv_cache`

File: models/demos/llama3/demo/simple_text_demo.py [hunk 11]
```diff
@@ -335,6 +439,19 @@ def test_llama_demo_text(
...
+
+    # uneven split of devices per DP group not supported
+    if data_parallel > num_devices or num_devices % data_parallel != 0:
+        pytest.skip(f"Invalid number of DP groups: {data_parallel}, for {num_devices} devices")
+
+    llama_dir = os.getenv("LLAMA_DIR")
+    if is_ci_env and num_devices == 32 and (data_parallel > 4 or (data_parallel == 4 and "3.1-70B" not in llama_dir)):
+        pytest.skip("CI runs only Llama3 70b DP = 4, TP = 8 on TG")
+    if is_ci_env and num_devices == 8 and data_parallel > 1 and not ("3.2-1B" in llama_dir or "3.1-8B" in llama_dir):
+        pytest.skip("CI runs only hybrid Llama3 1b and 8b on T3K")
```
[the remark this answers is not in the document]
Comment (ipotkonjak-tt) [reply models/demos/llama3/demo/simple_text_demo.py:470]: Wanted to avoid burdening the CI with additional tests. 1B and 8B seemed okay to cover perf regression checks as the smallest and biggest variants of the smaller Llama3 models. Should we add 3B anyway?

File: models/demos/llama3/demo/simple_vision_demo.py (hunk 3 above)
Comment (ipotkonjak-tt) [reply models/demos/llama3/demo/simple_vision_demo.py:123]: Wanted to avoid running additional tests by default, it can be uncommented if necessary

File: models/demos/llama3/demo/simple_text_demo.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (skhorasganiTT) [reply models/demos/llama3/demo/simple_text_demo.py:500]: Ok sure, I think those should be sufficient

File: models/demos/llama3/demo/simple_vision_demo.py (hunk 3 above)
Comment (skhorasganiTT) [reply models/demos/llama3/demo/simple_vision_demo.py:123]: Yeah I think there should be at least one DP test for the vision model

File: models/demos/llama3/demo/simple_text_demo.py [hunk 12]
```diff
@@ -155,7 +147,81 @@ def create_tt_model(
...
+        )
+    return page_table
+
+
+def prepare_generator_args(
+    num_devices,
+    data_parallel,
+    mesh_device,
+    instruct,
+    batch_size,
```
Comment (skhorasganiTT) [inline models/demos/llama3/demo/simple_text_demo.py]: A few remaining in `prepare_generator_args` and `create_tt_page_table` that should be `global_batch_size` instead of `max_batch_size` and `batch_size`

File: models/demos/llama3/tt/generator_vllm.py [hunk 13]
```diff
@@ -170,16 +238,20 @@ def prefill_forward(
...
+
 
 @INPUT_REGISTRY.register_input_processor(input_processor_for_llama_text)
 class TtLlamaForCausalLM(LlamaGenerator):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
     @classmethod
-    def initialize_vllm_model(cls, hf_config, mesh_device, max_batch_size, n_layers=None):
+    def initialize_vllm_model(cls, hf_config, tt_data_parallel, mesh_device, max_batch_size, n_layers=None):
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator_vllm.py]: For the `initialize_vllm_model` functions and the `allocate_kv_cache` function at the top, could you make `tt_data_parallel` a kw_arg with default 1 to preserve the existing functionality?

File: models/demos/llama3/tt/generator.py [hunk 14]
```diff
@@ -43,47 +43,70 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
+            user_id = group_user_id + model_id * batch_per_device
+
+            seq_len = prompt_lens[user_id]
+            last_token_idx = seq_len - 1
 
             # Since we give unpadded_seq_len, only the tile containing the last token is returned
-            output_logits[user_id] = logits
+            output_logits[user_id] = self.model[model_id].process_output_prefill(
+                out, last_token_idx=(last_token_idx % 32)
+            )
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator.py]: Was `process_output_prefill` moved out of the single user calls to delay the host readbacks? I'm wondering if this could cause us to run out of dram for longer seq lens since the logits are now being kept on device for all users. Maybe it should read out at the end of each DP loop instead?

Same comment for the vision model (`prefill_forward`)

File: models/tt_transformers/tt/generator.py [hunk 15]
```diff
@@ -464,12 +562,18 @@ def decode_forward(
...
+            return to_host
         else:
             return tt_logits
 
-    def read_decode_output(self, tt_logits, unpadded_batch, argmax_on_device=False):
-        logits = self.model.process_output_decode(tt_logits, B=unpadded_batch, S=1, argmax_on_device=argmax_on_device)
+    def read_decode_output(self, tt_logits, unpadded_batch, i, argmax_on_device=False):
+        logits = self.model[i].process_output_decode(
+            tt_logits, B=unpadded_batch, S=1, argmax_on_device=argmax_on_device
+        )
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py:576]: This function is used in vLLM as well, so the function signature should be kept the same (not pass in `i` and handle batch concat inside)

File: models/demos/llama3/demo/simple_text_demo.py (hunk 12 above)
Comment (ipotkonjak-tt) [reply models/demos/llama3/demo/simple_text_demo.py]: Renamed

File: models/demos/llama3/tt/generator_vllm.py (hunk 13 above)
Comment (ipotkonjak-tt) [reply models/demos/llama3/tt/generator_vllm.py]: Defaults to 1 now

File: models/tt_transformers/tt/generator.py (hunk 15 above)
Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py]: Thanks, yeah that should be fine

File: models/demos/llama3/tt/generator.py (hunk 14 above)
Comment (ipotkonjak-tt) [reply models/demos/llama3/tt/generator.py]: It was moved to reduce host / device transfers during prefill. Yes i guess we are at risk of using up all DRAM if we are to do it like this. We can move it to the end of outer loop and sync after prefill of N users, N being the number of submeshes. Currently we issue user prefill in a round robin manner across submeshes, so prefill of each "group" of users will be done in parallel

Comment (skhorasganiTT) [reply models/demos/llama3/tt/generator.py]: @ipotkonjak-tt reminded me that we're currently only slicing the output tile for prefill, so this is probably not an issue currently, unless we later scale to large batches or add tests (e.g. perplexity/acc evals) which require returning all the prefill logits. I think it might still be worth fixing for those scenarios, what do you think @yieldthought?

Comment (yieldthought) [reply models/demos/llama3/tt/generator.py]: We have plenty of DRAM. For a large model the dim of this is 32x8192 = 512KB per model. I'm not at all worried about running out of DRAM even with large batch sizes. If we do, we can add an environment variable to change this behaviour _then_. Let's not worry about it _now_ because it's probably not going to happen during the lifetime of this code and will be a trivial fix if it does.

File: models/tt_transformers/tt/common.py [hunk 16]
```diff
@@ -370,53 +382,19 @@ def sample_top_p(probs: torch.Tensor, p: float):
     return torch.gather(probs_idx, -1, next_token)
 
 
-def sample_host(tt_input, mesh_device, temperature=0.6, top_p=0.08, on_host=True):
+def sample_host(tt_input, temperature=0.6, top_p=0.08, on_host=True):
```
[the remark this answers is not in the document]
Comment (ipotkonjak-tt) [reply models/tt_transformers/tt/common.py:385]: `sample_host` is never actually called with `mesh_device` that isn't none. So the unused code was removed. In addition to this that code assumed TP implementation so it wasn't safe to keep it there with the new DP/HP configurations. The `ttnn.argmax` call is located in `ttnn_decode_forward` function of `Transformer` class, that remains unchanged

File: models/tt_transformers/PERF.md [hunk 17]
```diff
@@ -8,32 +8,37 @@ Also note that all the performance metrics below were taken for a maximum genera
...
-| Phi3.5-mini    | N300   |           |           | 57.8          | 62        |
-| Phi3.5-mini    | T3K    |           |           | 48.8          | 51        |
+This configuration uses bfp4 MLP FF1+FF3 for all models. **Batch_size=1 (per DP group) and prefill_length is 128 tokens.**
+
+| Model          | Device     | Top-1 (%) | Top-5 (%) | Speed (t/s/u) | TTFT (ms) |
+|----------------|------------|-----------|-----------|---------------|-----------|
+| Llama3.2-1B    | N150       | 88        | 98        | 84.5          | 58        |
+| Llama3.2-1B    | N300       | 91        | 98        | 100.5         | 54        |
+| Llama3.2-1B    | T3K        | 89        | 98        | 113.8         | 41        |
+| Llama3.2-1B    | T3K  (DP=4)|           |           | 83.5          | 19        |
```
[the remark this answers is not in the document]
Comment (mtairum) [reply models/tt_transformers/PERF.md:18]: Gotcha. Can you create an issue for the test_accuracy then, so we can add it in the future?

**Review by yieldthought — APPROVED** [review]
